crypto/ec/ec_mult.c

```diff
@@ -101,6 +101,225 @@ void EC_ec_pre_comp_free(EC_PRE_COMP *pre)
     OPENSSL_free(pre);
 }
 
+#define EC_POINT_BN_set_flags(P, flags) do { \
+    BN_set_flags((P)->X, (flags)); \
+    BN_set_flags((P)->Y, (flags)); \
+    BN_set_flags((P)->Z, (flags)); \
+} while(0)
+
+/*
+ * This functions computes (in constant time) a point multiplication over the
+ * EC group.
+ *
+ * At a high level, it is Montgomery ladder with conditional swaps.
+ *
+ * It performs either a fixed scalar point multiplication
+ *          (scalar * generator)
+ * when point is NULL, or a generic scalar point multiplication
+ *          (scalar * point)
+ * when point is not NULL.
+ *
+ * scalar should be in the range [0,n) otherwise all constant time bets are off.
+ *
+ * NB: This says nothing about EC_POINT_add and EC_POINT_dbl,
+ * which of course are not constant time themselves.
+ *
+ * The product is stored in r.
+ *
+ * Returns 1 on success, 0 otherwise.
+ */
+static int ec_mul_consttime(const EC_GROUP *group, EC_POINT *r, const BIGNUM *scalar,
+                            const EC_POINT *point, BN_CTX *ctx)
+{
+    int i, order_bits, group_top, kbit, pbit, Z_is_one;
+    EC_POINT *s = NULL;
+    BIGNUM *k = NULL;
+    BIGNUM *lambda = NULL;
+    BN_CTX *new_ctx = NULL;
+    int ret = 0;
+
+    if (ctx == NULL && (ctx = new_ctx = BN_CTX_secure_new()) == NULL)
```

Comment:
This seems ok to me. Probably it is not widely used because it is a very new API. Although I do wonder whether @richsalz has any view on this?

Your question does make me wonder though whether we should be calling BN_clear() on any of the intermediate temporary variables.

Reply:
So, @romen, do you think we need to call BN_clear() on the intermediate temporary vars?

Reply:
I would say it would be overkill and unnecessary as I don't think the threat model is considering malicious access to memory locations of running functions, what we need to avoid is leaving secret values in the heap after we release control of that memory area. 

We already preallocate words for those BN so that they don't need to be resized during computation, and shouldn't be automatically reduced during execution, so clearing them at the end when the CTX is released should be fine.

@bbbrumley do you agree?
@mattcaswell what are your thoughts on it?

Reply:
Ah! Right I missed the fact that BN_CTX_free() calls BN_clear_free() on those vars.